File: 0787-cheapest-flights-within-k-stops/0787-cheapest-flights-within-k-stops.py
class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, k: int) -> int:
        prices = [float("inf")]*n
        prices[src] = 0

        for _ in range(k+1):
            new_prices = prices.copy()
            for a,b,cost in flights:
                new_prices[b] = min(new_prices[b], prices[a]+cost)
            prices = new_prices
        
        return prices[dst] if prices[dst]<float("inf") else -1

File: 0787-cheapest-flights-within-k-stops/0787-cheapest-flights-within-k-stops.cpp
class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) {
        constexpr int infinity = numeric_limits<int>::max();

        vector<int> prices(n, infinity);
        auto new_prices = prices;
        prices[src] = 0;

        for (size_t i=0; i<k+1; ++i) {
            for (const auto& row : flights) {
                const size_t a = row[0];
                const size_t b = row[1];
                const int cost = row[2];
                if (prices[a]==infinity) {
                    continue;
                }

                new_prices[b] = std::min(new_prices[b], prices[a]+cost);
            }
            prices = new_prices;
        }

        if (prices[dst] == infinity) return -1;
        return prices[dst];
    }
};